File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/offline/OfflineRegionDefinition.java
package org.maplibre.nativeffi.offline;

import java.util.Objects;
import org.maplibre.nativeffi.geo.Geometry;
import org.maplibre.nativeffi.geo.LatLngBounds;

/** Offline region definition copied into native storage at creation time. */
public sealed interface OfflineRegionDefinition
    permits OfflineRegionDefinition.TilePyramid, OfflineRegionDefinition.GeometryRegion {
  record TilePyramid(
      String styleUrl,
      LatLngBounds bounds,
      double minZoom,
      double maxZoom,
      float pixelRatio,
      boolean includeIdeographs)
      implements OfflineRegionDefinition {
    public TilePyramid {
      Objects.requireNonNull(styleUrl, "styleUrl");
      Objects.requireNonNull(bounds, "bounds");
    }
  }

  record GeometryRegion(
      String styleUrl,
      Geometry geometry,
      double minZoom,
      double maxZoom,
      float pixelRatio,
      boolean includeIdeographs)
      implements OfflineRegionDefinition {
    public GeometryRegion {
      Objects.requireNonNull(styleUrl, "styleUrl");
      Objects.requireNonNull(geometry, "geometry");
    }
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/offline/OfflineRegionDefinition.kt
package org.maplibre.nativeffi.offline

import org.maplibre.nativeffi.geo.Geometry
import org.maplibre.nativeffi.geo.LatLngBounds

/** Offline region definition copied into native storage at creation time. */
public sealed interface OfflineRegionDefinition {
  public data class TilePyramid(
    public val styleUrl: String,
    public val bounds: LatLngBounds,
    public val minZoom: Double,
    public val maxZoom: Double,
    public val pixelRatio: Float,
    public val includeIdeographs: Boolean,
  ) : OfflineRegionDefinition

  public data class GeometryRegion(
    public val styleUrl: String,
    public val geometry: Geometry,
    public val minZoom: Double,
    public val maxZoom: Double,
    public val pixelRatio: Float,
    public val includeIdeographs: Boolean,
  ) : OfflineRegionDefinition
}
